README
Number::YAUID
=============

INSTALLATION

Just follow the usual procedure:

   perl Makefile.PL
   make
   make test
   make install


REPORTING BUGS

Alexander Borisov <lex.borisov@gmail.com>

COPYRIGHT

Copyright 2014-2016 Alexander Borisov

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Changes
1.60	Mon May 16 2016 12:50:19 GMT+0300
	Added new function: get_key_by_timestamp


1.50	Wed Oct 08 2014 17:11:25 GMT+0400
	+ Change limits: max node id 65535 => 16383, max inc id 32767 => 131071
	+ Add LIMIT_MIN_NODE_ID = 1

1.40	Wed Aug 06 2014 18:49:42 GMT+0400
	+ Add function get_period_key_by_timestamp and get_period_key_by_datetime

1.33	Wed Jul 30 2014 11:42:37 GMT+0400
	+ Fix pod SYNOPSIS :)

1.32	Wed Jul 30 2014 10:51:14 GMT+0400
	+ Add function timestamp_to_datetime
	+ Fix bugs

1.31	Tue Jul 29 2014 23:04:20 GMT+0400
	+ Release 1.31



YAUID.pm
package Number::YAUID;

use utf8;
use strict;
use vars qw($AUTOLOAD $VERSION $ABSTRACT @ISA @EXPORT);

BEGIN {
	$VERSION = 1.60;
	$ABSTRACT = "A decentralized unique ID generator (int64)";
	
	@ISA = qw(Exporter DynaLoader);
	@EXPORT = qw(
		get_error_text_by_code
		get_max_inc get_max_node_id get_max_timestamp
		timestamp_to_datetime
		get_key_by_timestamp
		get_period_key_by_timestamp get_period_key_by_datetime
	);
};

bootstrap Number::YAUID $VERSION;

use DynaLoader ();
use Exporter ();

my %opt = (
	try_count  => 0,
	sleep_usec => 35000,
	node_id    => 0
);

sub new {
	my ($class, $fuid_lock, $fnode_id) =
	   (shift, shift, shift);
	my %args = (
		%opt,
		@_
	);
	
	my $node_id = delete $args{node_id};
	my $self = $class->init($fuid_lock, $fnode_id);
	
	if(ref $self)
	{
		if(my $error = $self->get_error_code())
		{
			undef $self;
			return $error;
		}
		
		foreach my $pname (keys %args)
		{
			my $sub_name = "set_$pname";
			$self->$sub_name($args{$pname}) if exists $args{$pname};
		}
		
		unless($fnode_id || !ref $self)
		{
			$self->set_node_id($node_id);
		}
	}
	else {
		return YAUID_ERROR_CREATE_OBJECT();
	}
	
	$self;
}

sub timestamp_to_datetime {
	my @parts = (localtime( $_[0] || time ))[0..5];
	
	$parts[4] += 1;
	$parts[5] += 1900;
	
	sprintf("%04d-%02d-%02d %02d:%02d:%02d", reverse @parts);
}


1;


__END__

=head1 NAME

Number::YAUID - A decentralized unique ID generator (int64)

=head1 SYNOPSIS

 
 use Number::YAUID;
 
 my $object = Number::YAUID->new("/tmp/lock.file", "/etc/node.id");
 # OR
 #my $object = Number::YAUID->new("/tmp/lock.file", undef, node_id => 321);
 die get_error_text_by_code($object) unless ref $object;
 
 print "Max inc on sec: ", get_max_inc(), "\n";
 print "Max node id: ", get_max_node_id(), "\n";
 print "Max timestamp: ", get_max_timestamp(), "\n";
 
 foreach (0..5000)
 {
 	my $key = $object->get_key();
 	die get_error_text_by_code($object->get_error_code()) if $object->get_error_code();
 	
 	print "key: ", $key, "\n";
 	print "\ttimestamp: ", timestamp_to_datetime( $object->get_timestamp_by_key($key) ), "\n";
  	print "\tnode id: "  , $object->get_node_id_by_key($key)  , "\n";
 	print "\tinc id: "   , $object->get_inc_id_by_key($key)   , "\n";
 }
 

=head1 DESCRIPTION

Id generation at a node should not require coordination with other nodes. Ids should be roughly time-ordered when sorted lexicographically.

=head1 METHODS

=head2 new

 my $object = Number::YAUID->new(<file path to lockfile>, <file path to node id>[, params args]);

 my %p_args = (
 	try_count  => 0,     # count of key get attempts, 0 = unlimited
 	sleep_usec => 35000, # sleep 0.35 sec if limit key inc expired on current second
 	node_id    => 321    # current node id
 );
 
 my $object = Number::YAUID->new("/tmp/lock.file", undef, %p_args);
 
Create and prepare base structure. Return object or undef if something went wrong.

=head2 get_key

 my $key = $object->get_key();

Return a unique ID

=head2 get_period_key_by_datetime

 my $key = get_period_key_by_datetime(<from datetime>, <to datetime>, <from node ID>, <to node ID>);

=over 4

=item from datetime

YYYY-MM-DD hh:mm:ss

=item to datetime

YYYY-MM-DD hh:mm:ss if <to datetime> = 0, then <to datetime> = <from datetime>

=item from node ID

1 to get_max_node_id() if <from node ID> = 0, then <from node ID> = 1

=item to node ID

1 to get_max_node_id() if <to node ID> = 0, then <to node ID> = get_max_node_id()

=back

Return arrey ref where [0] = min, [1] = max unique ID

=head2 get_period_key_by_timestamp

 my $key = get_period_key_by_timestamp(<from timestamp>, <to timestamp>, <from node ID>, <to node ID>);

=over 4

=item from timestamp

timestamp

=item to timestamp

timestamp if <to timestamp> = 0, then <to timestamp> = <from timestamp>

=item from node ID

1 to get_max_node_id() if <from node ID> = 0, then <from node ID> = 1

=item to node ID

1 to get_max_node_id() if <to node ID> = 0, then <to node ID> = get_max_node_id()

=back

Return arrey ref where [0] = min, [1] = max unique ID

=head2 get_error_code

 $object->get_error_code();

Return error code.

=head2 get_error_text_by_code

 get_error_text_by_code(<error code>);

Return description by error code

=head2 get_timestamp_by_key

 $object->get_timestamp_by_key(<key>);

Return timestamp from a key

=head2 get_node_id_by_key

 $object->get_node_id_by_key(<key>);

Return node id from a key

=head2 get_inc_id_by_key

 $object->get_inc_id_by_key(<key>);

Return inc id from a key

=head2 timestamp_to_datetime

 timestamp_to_datetime(<timestamp>);

Convert timestamp to datetime (YYYY-MM-DD hh:mm:ss)

=head1 DESTROY

 undef $obj;

Free mem and destroy object.

=head1 AUTHOR

Alexander Borisov <lex.borisov@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Alexander Borisov.

This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=cut


yauid/yauid.h
/*
 Copyright (c) 2014-2016 Alexander Borisov
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#ifndef yauid_yauid_h
#define yauid_yauid_h

#ifdef __cplusplus
extern "C" {
#endif
    
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <inttypes.h>
#include <time.h>
#include <sys/file.h>
#include <unistd.h>
    
#if defined(__x86_64__) || defined(__ppc64__) || defined(_WIN64)
#define ENVIRONMENT64
#else
#define ENVIRONMENT32
#endif
    
    /***********************************************************************************
     *
     * Base init
     *
     ***********************************************************************************/
    
#ifndef LIMIT_MIN_NODE_ID
#define LIMIT_MIN_NODE_ID    1L
#endif
    
#ifndef BIT_LIMIT
#define BIT_LIMIT           64L
#endif
    
#ifndef BIT_LIMIT_TIMESTAMP
#define BIT_LIMIT_TIMESTAMP 33L
#endif
    
#ifndef BIT_LIMIT_NODE
#define BIT_LIMIT_NODE      14L
#endif
    
#define BIT_LIMIT_INC       (BIT_LIMIT - (BIT_LIMIT_TIMESTAMP + BIT_LIMIT_NODE))
    
#define NUMBER_LIMIT           ((1L << BIT_LIMIT_INC) - 1)
#define NUMBER_LIMIT_NODE      ((1L << BIT_LIMIT_NODE) - 1)
#define NUMBER_LIMIT_TIMESTAMP ((1L << BIT_LIMIT_TIMESTAMP) - 1)
    
    // 64 bit
    typedef uint64_t hkey_t;
    
    enum yauid_status {
        YAUID_OK = 0,
        YAUID_ERROR_CREATE_KEY_FILE,
        YAUID_ERROR_OPEN_LOCK_FILE,
        YAUID_ERROR_KEYS_ENDED,
        YAUID_ERROR_FILE_NODE_ID,
        YAUID_ERROR_FILE_NODE_MEM,
        YAUID_ERROR_FILE_NODE_EXT,
        YAUID_ERROR_FILE_LOCK,
        YAUID_ERROR_LONG_NODE_ID,
        YAUID_ERROR_SHORT_NODE_ID,
        YAUID_ERROR_READ_KEY,
        YAUID_ERROR_FILE_SEEK,
        YAUID_ERROR_WRITE_KEY,
        YAUID_ERROR_FLUSH_KEY,
        YAUID_ERROR_TRY_COUNT_KEY,
        YAUID_ERROR_CREATE_OBJECT,
        YAUID_ERROR_ALLOC_KEY_FILE
    }
    typedef yauid_status_t;
    
    // base structure
    struct yauid {
        int           i_lockfile;
        char*         c_lockfile;
        FILE*         h_lockfile;
        unsigned long node_id;
        
        unsigned int try_count;
        useconds_t sleep_usec;
        
        enum yauid_status error;
        void *ext_value;
    }
    typedef yauid;
    
    struct yauid_period_key {
        hkey_t min;
        hkey_t max;
    }
    typedef yauid_period_key;
    
    /***********************************************************************************
     *
     * YAUID
     *
     ***********************************************************************************/
    
    /**
     * Create a new yauid
     *
     * @param[in] File path to lock file. Important! All yauid (on one node) link to this file
     * @param[in] NULL or file path to node id. See yauid_set_node_id
     * @return yauid structure
     */
    yauid * yauid_init(const char *filepath_key, const char *filepath_node_id);
    
    /**
     * Frees all allocated resources
     *
     * @param[in] yauid
     */
    void yauid_destroy(yauid* yaobj);
    
    /**
     * Get unique key by current node
     *
     * @param[in] yauid
     * @return unique key if successful or 0 if any error. See yauid_get_error_code
     */
    hkey_t yauid_get_key(yauid* yaobj);
    
    /**
     * Tries once get unique key by current node
     *
     * @param[in] yauid
     * @return unique key if successful or 0 if any error. See yauid_get_error_code
     */
    hkey_t yauid_get_key_once(yauid* yaobj);
    
    /**
     * Set current node id
     *
     * @param[in] yauid
     * @param[in] node id equal yauid_get_min_node_id() to yauid_get_max_node_id()
     */
    void yauid_set_node_id(yauid* yaobj, unsigned long node_id);
    
    /**
     * Set sleeping time to try get key (refers to yauid_get_key)
     *
     * @param[in] yauid
     * @param[in] microsecond intervals. Default: 35000 (35000 microsecond = 0.3 second)
     */
    void yauid_set_sleep_usec(yauid* yaobj, useconds_t sleep_usec);
    
    /**
     * Number of attempts before surrender (refers to yauid_get_key)
     *
     * @param[in] yauid
     * @param[in] try count. Default: 0 (0 is infinity)
     */
    void yauid_set_try_count(yauid* yaobj, unsigned int try_count);
    
    /**
     * Get timestamp from yauid key
     *
     * @param[in] yauid key
     * @return timestamp
     */
    unsigned long yauid_get_timestamp(hkey_t key);
    
    /**
     * Get node id from yauid key
     *
     * @param[in] yauid key
     * @return node id
     */
    unsigned long yauid_get_node_id(hkey_t key);
    
    /**
     * Get inc (unique id per second in the current key) id from yauid key
     *
     * @param[in] yauid key
     * @return inc id
     */
    unsigned long yauid_get_inc_id(hkey_t key);
    
    /**
     * Get the maximum possible inc per one second
     *
     * @return inc id
     */
    unsigned long long int yauid_get_max_inc();
    
    /**
     * Get the minimum possible node id
     *
     * @return node id
     */
    unsigned long long int yauid_get_min_node_id();
    
    /**
     * Get the maximum possible node id
     *
     * @return node id
     */
    unsigned long long int yauid_get_max_node_id();
    
    /**
     * Get the maximum possible timestamp
     *
     * @return timestamp
     */
    unsigned long long int yauid_get_max_timestamp();
    
    /**
     * Convert datetime to timestamp. 2014-07-12 04:23:12 => 1405124592
     *
     * @param[in] datetime (e.g. 2014-07-12 04:23:12)
     * @return timestamp
     */
    time_t yauid_datetime_to_timestamp(const char *datetime);
    
    /**
     * Get YAUID by timestamp
     *
     * @param[in] timestamp like a 1405124592
     * @param[in] node id; > 0 and <= NUMBER_LIMIT_NODE
     * @param[in] increment counter; >= 0 and <= NUMBER_LIMIT
     *
     * @return unique key if successful or 0 if limits are exceeded
     */
    hkey_t
    yauid_get_key_by_timestamp(time_t timestamp, size_t node_id, size_t counter);
    
    /**
     * Get minimum and maximum key from date and node id by timestamp
     *
     * @param[in] from timestamp (e.g. 1405124592)
     * @param[in] to timestamp. NULL = from timestamp
     * @param[in] from node id
     * @param[in] to node id. NULL = from node
     * @param[out] min and max yauid key
     */
    void yauid_get_period_key_by_timestamp(time_t from_timestamp,
                                           time_t to_timestamp,
                                           unsigned long long int from_node_id,
                                           unsigned long long int to_node_id,
                                           yauid_period_key *pkey);
    
    /**
     * Get minimum and maximum key from date and node id by datetime
     *
     * @param[in] from datetime (e.g. "2014-07-12 04:23:12")
     * @param[in] to datetime. NULL = from datetime
     * @param[in] from node id
     * @param[in] to node id. NULL = from node
     * @param[out] min and max yauid key
     */
    void yauid_get_period_key_by_datetime(const char *from_datetime,
                                          const char *to_datetime,
                                          unsigned long long int from_node_id,
                                          unsigned long long int to_node_id,
                                          yauid_period_key *pkey);
    
    /**
     * Get error description by error code
     *
     * @param[in] error code. See yauid_get_error_code
     * @return pointer to description
     */
    char * yauid_get_error_text_by_code(yauid_status_t error_code);
    
    /**
     * Get current status of yauid
     *
     * @param[in] yauid
     * @return status code from enum yauid_status
     */
    yauid_status_t yauid_get_error_code(yauid* yaobj);
    
    
    
#ifdef __cplusplus
} /* extern "C" */
#endif

#endif

